### ISSUE TITLE
ncm-ncd 16.6.0-rc3: cannot acquire ncm-ncd lock during upgrade

### ISSUE THREAD

**Issue by jouvin** [quattor/ncm-ncd#69]
During Quattor client upgrade through `ncm-cdispd`, `ncd-ncd` is run 3 times: this is due to `ncm-cdispd` service restart during upgrade of `ncm-cdispd` and `cdp-listend`. Second and third `ncm-ncd` instances are blocked by `ncm-ncd` lock, until the previous instance has completed.

It seems that the new `CAF::Lock` in 16.6.0-rc3 causes the third instance to exit before running with the message:
```
2016/07/11-20:16:41 [ERROR] cannot acquire lock after 10 tries: /var/lock/quattor/ncm-ncd
```

I don't think this was the case before (it was just waiting until the process whose pid was in the lock file had exited). It should not be a problem (in fact only one restart of `ncm-cdispd` is normally enough) but if for some reasons `ncm-ncd` takes longer to complete, there is may the risk that it will not be restarted (while the running one will be terminated at the end of `ncm-ncd` run), leaving the system unmanaged.

Is it possible to confirm that this is not a problem or to ensure a longer (infinite?) wait?

Probably a side effect of #64. May be good to fix it or revert #64 before releasing 16.6.0-rc3, even if not strictly speaking a showstopper...

Would be good to look at this issue before releasing [16.6.0-rc3](https://github.com/quattor/release/issues/250), even if not strictly speaking a showstopper...

**Comment by stdweird** [comment]
@jouvin what is the timeout? (i'll open a pr to add it to the message and add unittests). that part of the code in CAF looks the same, and in ncm-ncd it is not modified.
#64 only add logging, but even without it, it should have displayed the error i think

**Comment by stdweird** [comment]
@jouvin https://github.com/quattor/CAF/pull/175
looks like one retry was missing, but seems to work otherwise (and the old code typically slept half less).
is there a way to check old logs to see if during previous upgrades that triggered the 3 ncm-ncd, the 3rd didn't stop with an error?

**Comment by jouvin** [comment]
I will check if I find evidence in logs. But the reason I identified the problem is that there were only 2 entries in `ncm-spma` log file created during the upgrade, instead of 3 in previous versions... SO I'm sure the 3d instance was running succesfully.

**Comment by stdweird** [comment]
if the spma logs have timestamnps, can you see if the retries * timeout for ncm-ncd was high enough? i'll do some testing later today

**Comment by jouvin** [comment]
I did a new test today and unfortunately was not able to reproduce the problem... Usual behaviour with this kind of timing-related condition. The log even shows that it works very much as expected!
```
Jul 15 15:40:15 grid63 /usr/sbin/cdp-listend[3692881]: Received UDP packet (ccm|1468590007) from 134.158.88.149
Jul 15 15:40:15 grid63 /usr/sbin/cdp-listend[3692881]: /usr/sbin/ccm-fetch will be called in 23 seconds (offset=0, smear=23)
Jul 15 15:40:38 grid63 /usr/sbin/cdp-listend[3692881]: Calling /usr/sbin/ccm-fetch with unix time 1468590007 (after 23 seconds)
Jul 15 15:40:54 grid63 ncm-cdispd[3692747]: new profile arrived, examining...
Jul 15 15:40:54 grid63 ncm-cdispd[3692747]: new (and changed) profile detected
Jul 15 15:40:56 grid63 ncd[3701482]: NCM-NCD version 16.2.0 started by root at: Fri Jul 15 15:40:56 2016
Jul 15 15:40:56 grid63 ncd[3701482]: executing configure on components....
Jul 15 15:42:24 grid63 ncm-cdispd[3692747]: Signal TERM registered for processing after ncm-ncd completion
Jul 15 15:42:25 grid63 ncm-cdispd[3702388]: ncm-cdispd version 16.2.0 started by root at: Fri Jul 15 15:42:25 2016 pid: 3702388
Jul 15 15:42:25 grid63 ncm-cdispd[3702388]: initalization of components
Jul 15 15:42:26 grid63 ncd[3702390]: NCM-NCD version 16.2.0 started by root at: Fri Jul 15 15:42:26 2016
Jul 15 15:42:26 grid63 ncd[3702390]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:42:27 grid63 yum[3701497]: Updated: ncm-ncd-16.6.0-rc3_1.noarch
Jul 15 15:42:52 grid63 ncm-cdispd[3702388]: Signal TERM registered for processing after ncm-ncd completion
Jul 15 15:42:53 grid63 ncm-cdispd[3703249]: ncm-cdispd version 16.6.0-rc3 started by root at: Fri Jul 15 15:42:53 2016 pid: 3703249
Jul 15 15:42:53 grid63 ncm-cdispd[3703249]: initalization of components
Jul 15 15:42:53 grid63 yum[3701497]: Updated: ncm-cdispd-16.6.0-rc3_1.noarch
Jul 15 15:42:54 grid63 ncd[3703251]: NCM-NCD version 16.6.0-rc3 started by root at: Fri Jul 15 15:42:54 2016
Jul 15 15:42:54 grid63 ncd[3703251]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:42:56 grid63 /usr/sbin/cdp-listend[3703273]: Starting /usr/sbin/cdp-listend with configuration file /etc/cdp-listend.conf
Jul 15 15:42:56 grid63 /usr/sbin/cdp-listend[3703276]: /usr/sbin/cdp-listend daemon successfully started on port 7777
Jul 15 15:42:56 grid63 /usr/sbin/cdp-listend[3703276]: Startup invocation of /usr/sbin/ccm-fetch
Jul 15 15:42:56 grid63 yum[3701497]: Updated: cdp-listend-16.6.0-rc3_1.noarch
Jul 15 15:42:56 grid63 ncd[3702390]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:43:24 grid63 ncd[3703251]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:43:26 grid63 ncd[3702390]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:43:54 grid63 ncd[3703251]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:43:56 grid63 ncd[3702390]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:44:24 grid63 ncd[3703251]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:44:26 grid63 ncd[3702390]: Old style lock found /var/lock/quattor/ncm-ncd
Jul 15 15:44:35 grid63 ncd[3701482]: 0 errors, 2 warnings executing configure
Jul 15 15:44:35 grid63 ncm-cdispd[3692747]: ncm-ncd finished with status: 0 (ec 0, all configuration modules ran successfully)
Jul 15 15:44:35 grid63 ncm-cdispd[3692747]: Processing delayed signal TERM
Jul 15 15:44:35 grid63 ncm-cdispd[3692747]: signal handler: signal TERM received
Jul 15 15:44:35 grid63 ncm-cdispd[3692747]: terminating ncm-cdispd...
Jul 15 15:44:54 grid63 ncd[3703251]: executing configure on components....
Jul 15 15:46:51 grid63 ncd[3703251]: 0 errors, 2 warnings executing configure
Jul 15 15:46:51 grid63 ncm-cdispd[3703249]: ncm-ncd finished with status: 0 (ec 0, all configuration modules ran successfully)
Jul 15 15:46:56 grid63 ncd[3702390]: executing configure on components....
Jul 15 15:48:32 grid63 ncd[3702390]: 0 errors, 2 warnings executing configure
Jul 15 15:48:32 grid63 ncm-cdispd[3702388]: ncm-ncd finished with status: 0 (ec 0, all configuration modules ran successfully)
Jul 15 15:48:32 grid63 ncm-cdispd[3702388]: Processing delayed signal TERM
Jul 15 15:48:32 grid63 ncm-cdispd[3702388]: signal handler: signal TERM received
Jul 15 15:48:32 grid63 ncm-cdispd[3702388]: terminating ncm-cdispd...
```

This was obtained with the following command applied to the [full log](https://github.com/quattor/ncm-ncd/files/366099/upgrade.log.gz):
```
egrep 'ncd|version|cdispd|listend' /tmp/upgrade.log |grep -v globus|grep -v 'running component'
```

The process that most waited (`3702390`) had only 5 retries... So I don't know what caused the 10 limit to be reached during my first attempt... May be more changes to apply in `ncm-spma` (I may have updated the YUM snapshot used for the OS at the same time. The retry interval seems to be 30s, 10 retries mean 5 mn which may be short if there are a lot of changes to be applied by YUM... If trivial, I would consider doubling the limit at least...

**Comment by stdweird** [comment]
yes, the limit seems low (but old lock code only did half of that). 5 minutes might certainly not be enough for major OS update (esp in a cluster with lots of nodes pulling in rpms). but we also run the defaults.
i'm going to guess that the new logging shows the error, and the old one not at all or not in the same logfile.

**Comment by jouvin** [comment]
Note that the low limit is a problem only if you upgrade Quattor client at the same time you do another upgrade: this is normally the only place where you will have to run `ncm-ncd` several time (from different `ncm-cdispd`) and you don't want `ncm-cdispd` to be killed because `ncm-ncd` was not able to run due to the lock (which is what happens if I'm right).

**Comment by jouvin** [comment]
Any plan to do something for this potential problem or do we say that we live with it for the time being and postpone to a later release?

**Comment by jouvin** [comment]
As @stdweird is not available to work more on it in the short term, I propose to postpone this issue and review after next RC if we need it or not.

**Comment by jrha** [comment]
Any thoughts on whether we need this now that we have rc5?

**Comment by stdweird** [comment]
need what? higher defaults? sure, why not.

**Comment by jouvin** [comment]
Based on the deployment at LAL, I'd say nothing urgent... We can release 16.6.0 with this value which is already higher than before and plan for 16.8 a further increase...

**Comment by jrha** [comment]
Okay!

**Comment by jrha** [comment]
Agreed investigate further at standup meeting today.

**Comment by jouvin** [comment]
To be a little bit more specific, there is basic agreement that a longer timeout in `ncm-ncd` (10 to 15 mn) should not hurt... but it would be good to clarify the exact impact of `ncm-ncd` failing to acquire the lock on `ncm-ncd`. When I saw the problem, I got the feeling (evidence?) that `ncm-cdispd` was terminated because of this. Normally a `ncm-ncd` problem should never terminate `ncm-cdispd`, except may be during the initial run of `ncm-ncd` and this may explain why `ncm-cdispd` got terminated during the upgrade process (where the failure to acquire a lock happens during the initial phase of the new `ncm-cdispd` started by the RPM upgrade script).

**Comment by stdweird** [comment]
should be easy to test: take the lock on command line , and restart ncm-cdispd

```bash
PERL5LIB=/usr/lib/perl perl -e 'use CAF::Lock; my $l = CAF::Lock->new("/var/lock/quattor/ncm-ncd"); $l->set_lock(); sleep 1000'
```

`ncm-cdispd` stays running

extract from ncm-cdispd log

```
2016/08/18-21:18:37 [VERB] [VERB] accessing CCM cache manager..
2016/08/18-21:18:37 [VERB] [VERB] getting locked CCM configuration..
2016/08/18-21:18:37 [VERB] [VERB] checking for ncm-ncd locks...
2016/08/18-21:18:37 [VERB] [DEBUG] Empty lockfile /var/lock/quattor/ncm-ncd found: no old-style lock
2016/08/18-21:18:37 [VERB] [DEBUG] flock on /var/lock/quattor/ncm-ncd gave has_lock 0
...
2016/08/18-21:23:07 [VERB] [VERB] lock file is already held, try 10 out of 10 (timeout 30)
2016/08/18-21:23:37 [VERB] [DEBUG] Empty lockfile /var/lock/quattor/ncm-ncd found: no old-style lock
2016/08/18-21:23:37 [VERB] [DEBUG] flock on /var/lock/quattor/ncm-ncd gave has_lock 0
2016/08/18-21:23:37 [VERB] [ERROR] cannot acquire lock after 11 tries (timeout 30): /var/lock/quattor/ncm-ncd
...
2016/08/18-21:23:37 [WARN] ncm-ncd finished with status: 255 (ec 65280, some configuration modules failed to run successfully)
2016/08/18-21:23:37 [DEBUG] process_signal(): no delayed signal to process
2016/08/18-21:23:37 [DEBUG] Initializing $last_ncd_status to init_components status (1)
2016/08/18-21:23:37 [DEBUG] CID of reference configuration set to 49
2016/08/18-21:23:37 [DEBUG] checking for new profiles ...
2016/08/18-21:23:37 [DEBUG] CID of last profile processed: 49
2016/08/18-21:23:37 [VERB] no new profile found, sleeping
2016/08/18-21:23:37 [DEBUG] sleep for 60 seconds
...
```

**Comment by jouvin** [comment]
I'll try to double check but your test seems to demonstrate that I mixed up when reading log message `ncm-cdispd` and `ncm-ncd` termination....
